pkg/provider/kubernetes/gateway/kubernetes.go

```diff
@@ -1226,20 +1233,36 @@ func hostSNIRule(hostnames []v1alpha2.Hostname) (string, error) {
 
 		h := string(hostname)
 
-		// TODO support wildcard hostnames with an HostSNI regexp matcher
-		if strings.Contains(h, "*") {
-			return "", fmt.Errorf("wildcard hostname is not supported: %q", h)
+		wildcard := strings.Count(h, "*")
+		switch wildcard {
+		case 0:
+			matchers["HostSNI("] = append(matchers["HostSNI("], "`"+h+"`")
+		case 1:
+			if strings.HasPrefix(h, "*.") {
+				matchers["HostSNIRegexp("] = append(matchers["HostSNIRegexp("], "`"+strings.Replace(h, "*.", "{subdomain:[a-zA-Z0-9-]+}.", 1)+"`")
```

Comment:
If #9531 is merged, it should be the following:

```suggestion
				matchers["HostSNIRegexp("] = append(matchers["HostSNIRegexp("], "`"+strings.Replace(regexp.QuoteMeta(h), `\*\.`, `[a-zA-Z0-9-]+\.`, 1)+"`")
```